_Student: Kiseleva Elena_

In [2]:
# display all outputs, not only last one
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
print('done')

done


In [3]:
%load_ext line_profiler
%load_ext memory_profiler

### 1. Чтение/запись в словарь. При попытке чтения несуществующего ключа возвращается None.


__Запись в словарь последовательных данных.__

In [ ]:
# %load ./dictionary/write_to_dict.py 
import random


#prepare data
RANGE = 100000
ABC = 'abcdefghijklmnopqrstufwxyzABCDEFGHIJKLMNOPQRSTUFWXYZ'

inline_keys = [x for x in range(RANGE)]
inline_values = [''.join(random.sample(ABC, 8)) for _ in range(RANGE)]
random_keys = [random.randint(0, RANGE) for _ in range(RANGE)]
random_values = [''.join(random.sample(ABC, random.randint(8, 24))) for _ in range(RANGE)]


def write_to_dict_v1(keys, values):
    """
    Write items in dict with for-loop.
    The existing values will be overwritten.
    """
    result = {}
    for key, value in zip(keys, values):
        result[key] = value
    return result


def write_to_dict_v2(keys, values):
    """
    Write items in dictionary with method setdefault. 
    The existing items will not be overwritten.
    """
    result = {}
    for key, value in zip(keys, values):
        result.setdefault(key, value)
    return result


def write_to_dict_v3(keys, values):
    """
    Write items in dictionary with checking if the item just in the dictionary. 
    The existing items will not be overwritten.
    """
    result = {}
    for key, value in zip(keys, values):
        if key not in result:
            result[key] = value
    return result


def write_to_dict_v4(keys, values):
    """
    Write items in dictionary with dictionary comprehention. 
    The existing items will be overwritten.
    """
    return {key: value for key, value in zip(keys, values)}


In [ ]:
# %load ./dictionary/test_write_to_dict.py
import pytest

from write_to_dict import write_to_dict_v1, write_to_dict_v2, write_to_dict_v3, write_to_dict_v4


keys = [1, 2, 3, 4, 2]
values = ['abc', 'def', 'ghj', 'ikl', 'mno']
expected = {1: 'abc', 2: 'def', 3: 'ghj', 4: 'ikl'}

def test_write_to_dict_v1():
    actual = write_to_dict_v1(keys, values)
    expected = {1: 'abc', 2: 'mno', 3: 'ghj', 4: 'ikl'}
    assert actual == expected


def test_write_to_dict_v2():
    actual = write_to_dict_v2(keys, values)
    expected = {1: 'abc', 2: 'def', 3: 'ghj', 4: 'ikl'}
    assert actual == expected

    
def test_write_to_dict_v3():
    actual = write_to_dict_v3(keys, values)
    expected = {1: 'abc', 2: 'def', 3: 'ghj', 4: 'ikl'}
    assert actual == expected


def test_write_to_dict_v4():
    actual = write_to_dict_v4(keys, values)
    expected = {1: 'abc', 2: 'mno', 3: 'ghj', 4: 'ikl'}
    assert actual == expected


In [6]:
!python -m pytest ./dictionary/test_write_to_dict.py

============================= test session starts =============================
platform win32 -- Python 3.7.4, pytest-5.4.3, py-1.9.0, pluggy-0.13.1
rootdir: F:\!EPAM\HW_Lesson12
collected 4 items

dictionary\test_write_to_dict.py ....                                    [100%]

============================= 4 passed in 10.29s ==============================


In [7]:
%timeit write_to_dict_v1(inline_keys, inline_values)

30.2 ms ± 3.87 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [8]:
%timeit write_to_dict_v2(inline_keys, inline_values)

42 ms ± 4.72 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [9]:
%timeit write_to_dict_v3(inline_keys, inline_values)

50.1 ms ± 11.7 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [10]:
%timeit write_to_dict_v4(inline_keys, inline_values)

31.8 ms ± 7.55 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [11]:
import cProfile


def benchmark_write_inline(n):
    for i in range(n):
        write_to_dict_v1(inline_keys, inline_values)
        write_to_dict_v2(inline_keys, inline_values)
        write_to_dict_v3(inline_keys, inline_values)
        write_to_dict_v4(inline_keys, inline_values)


cProfile.run('benchmark_write_inline(100)', sort='tottime')

         10000504 function calls in 18.607 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      100    4.159    0.042    4.159    0.042 <ipython-input-4-fc9ecef60aeb>:37(write_to_dict_v3)
      100    4.014    0.040    7.010    0.070 <ipython-input-4-fc9ecef60aeb>:26(write_to_dict_v2)
      100    3.341    0.033    3.341    0.033 <ipython-input-4-fc9ecef60aeb>:15(write_to_dict_v1)
 10000000    2.996    0.000    2.996    0.000 {method 'setdefault' of 'dict' objects}
      100    2.989    0.030    2.989    0.030 <ipython-input-4-fc9ecef60aeb>:54(<dictcomp>)
        1    1.106    1.106   18.606   18.606 <ipython-input-11-1b277bfdd52f>:4(benchmark_write_inline)
      100    0.001    0.000    2.991    0.030 <ipython-input-4-fc9ecef60aeb>:49(write_to_dict_v4)
        1    0.000    0.000   18.607   18.607 {built-in method builtins.exec}
        1    0.000    0.000   18.606   18.606 <string>:1(<module>)
        1    0.000    0.000

In [64]:
%lprun -f benchmark_write_inline benchmark_write_inline(100)

Timer unit: 1e-07 s

Total time: 47.3423 s
File: <ipython-input-11-1b277bfdd52f>
Function: benchmark_write_inline at line 4

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     4                                           def benchmark_write_inline(n):
     5       101       9596.0     95.0      0.0      for i in range(n):
     6       100  112004194.0 1120041.9     23.7          write_to_dict_v1(inline_keys, inline_values)
     7       100  133569182.0 1335691.8     28.2          write_to_dict_v2(inline_keys, inline_values)
     8       100  160675949.0 1606759.5     33.9          write_to_dict_v3(inline_keys, inline_values)
     9       100   67164054.0 671640.5     14.2          write_to_dict_v4(inline_keys, inline_values)

In [62]:
# %load write_to_dict_memory_profiling.py
from memory_profiler import profile

from dictionary.write_to_dict import write_to_dict_v1, write_to_dict_v2, write_to_dict_v3, write_to_dict_v4
from dictionary.write_to_dict import inline_keys, inline_values


@profile
def run_all():
    write_to_dict_v1(inline_keys, inline_values)
    write_to_dict_v2(inline_keys, inline_values)
    write_to_dict_v3(inline_keys, inline_values)
    write_to_dict_v4(inline_keys, inline_values)


if __name__ == '__main__':
    run_all()


ERROR: Could not find file <ipython-input-62-9fab162994e8>
NOTE: %mprun can only be used on functions defined in physical files, and not in the IPython environment.


In [63]:
!python write_to_dict_memory_profiling.py

Filename: write_to_dict_memory_profiling.py

Line #    Mem usage    Increment   Line Contents
     7     27.9 MiB     27.9 MiB   @profile
     8                             def run_all():
     9     27.5 MiB      0.0 MiB       write_to_dict_v1(inline_keys, inline_values)
    10     27.5 MiB      0.0 MiB       write_to_dict_v2(inline_keys, inline_values)
    11     27.5 MiB      0.0 MiB       write_to_dict_v3(inline_keys, inline_values)
    12     27.5 MiB      0.0 MiB       write_to_dict_v4(inline_keys, inline_values)




Использование dictionary comprehension дает самый лучший результат (могу предположить, что это из-за внутренней реальзации данного средства языка). Вариант с прямой записью, без проверок на вхождение ключа в словарь, работает чуть быстрее остальных двух, так как в них тратится время на дополнительную проверку наличия ключа в словаре. 
Потерь по памяти нет, так как в словаре значения хранятся по ссылке.

__Запись в словарь рандомных данных.__

In [15]:
%timeit write_to_dict_v1(random_keys, random_values)

39.3 ms ± 4.59 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [16]:
%timeit write_to_dict_v2(random_keys, random_values)

58.5 ms ± 10.5 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [17]:
%timeit write_to_dict_v3(random_keys, random_values)

49.4 ms ± 9.87 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [18]:
%timeit write_to_dict_v4(random_keys, random_values)

39.9 ms ± 10.6 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [19]:
import cProfile


def benchmark_write_random(n):
    for i in range(n):
        write_to_dict_v1(random_keys, random_values)
        write_to_dict_v2(random_keys, random_values)
        write_to_dict_v3(random_keys, random_values)
        write_to_dict_v4(random_keys, random_values)


cProfile.run('benchmark_write_random(100)', sort='tottime')

         10000504 function calls in 20.181 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      100    4.276    0.043    4.276    0.043 write_to_dict.py:36(write_to_dict_v3)
      100    3.877    0.039    3.877    0.039 write_to_dict.py:14(write_to_dict_v1)
 10000000    3.806    0.000    3.806    0.000 {method 'setdefault' of 'dict' objects}
      100    3.768    0.038    7.574    0.076 write_to_dict.py:25(write_to_dict_v2)
      100    3.634    0.036    3.634    0.036 write_to_dict.py:53(<dictcomp>)
        1    0.818    0.818   20.181   20.181 <ipython-input-19-434c65d1593e>:4(benchmark_write_random)
      100    0.001    0.000    3.635    0.036 write_to_dict.py:48(write_to_dict_v4)
        1    0.000    0.000   20.181   20.181 {built-in method builtins.exec}
        1    0.000    0.000   20.181   20.181 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




In [20]:
%lprun -f benchmark_write_random benchmark_write_random(100)

Timer unit: 1e-07 s

Total time: 44.1011 s
File: <ipython-input-111-434c65d1593e>
Function: benchmark_write_random at line 4

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     4                                           def benchmark_write_random(n):
     5       101       7068.0     70.0      0.0      for i in range(n):
     6       100  109968005.0 1099680.1     24.9          write_to_dict_v1(random_keys, random_values)
     7       100  129207820.0 1292078.2     29.3          write_to_dict_v2(random_keys, random_values)
     8       100  131194029.0 1311940.3     29.7          write_to_dict_v3(random_keys, random_values)
     9       100   70633732.0 706337.3     16.0          write_to_dict_v4(random_keys, random_values)

При записи случайных ключей с повторами вариант с перезаписью ключей (v1) теряет свое преимущество, так как в данном случае фактически записываются все значения (с перезаписью), в то время как два других варианта повторяющиеся ключи не перезаписывают. Самым лучшим вариантом остается dictionary comprehension, хотя он тоже частично теряет по отношению к вариантам с проверками наличия ключа в словаре.

__Чтение данных из словаря.__

In [ ]:
# %load ./dictionary/read_from_dict.py 
import random


#prepare data
RANGE = 10000
ABC = 'abcdefghijklmnopqrstufwxyzABCDEFGHIJKLMNOPQRSTUFWXYZ'

inline_keys = [x for x in range(RANGE)]
inline_values = [''.join(random.sample(ABC, 8)) for _ in range(RANGE)]
random_keys = [random.randint(0, RANGE) for _ in range(RANGE)]
random_values = [''.join(random.sample(ABC, random.randint(8, 24))) for _ in range(RANGE)]

data = {k: v for k, v in zip(inline_keys, inline_values)}
data_random = {k: v for k, v in zip(random_keys, random_values)}


def read_from_dict_v1(dictionary, key):
    """
    Read items from the dictionary with method get.
    Return None if the key is not in dictionary.
    """
    return dictionary.get(key)


def read_from_dict_v2(dictionary, key):
    """
    Read items from the dictionary with checking if the key is in dictionary.
    Return None if the key is not in dictionary.
    """
    return dictionary[key] if key in dictionary else None


def read_from_dict_v3(dictionary, key):
    """
    Read items from the dictionary with interrupt handling.
    Return None if the key is not in dictionary.
    """
    try:
        value = dictionary[key]
    except KeyError:
        value = None
    return value


In [ ]:
# %load ./dictionary/test_read_from_dict.py
import pytest

from read_from_dict import read_from_dict_v1, read_from_dict_v2, read_from_dict_v3


test_data = {1: 'abc', 4: 'def', 6: 'hjik'}


def test_in_dict_v1():
    assert read_from_dict_v1(test_data, 4) == 'def'
    

def test_not_in_dict_v1():
    assert read_from_dict_v1(test_data, 2) is None


def test_in_dict_v2():
    assert read_from_dict_v1(test_data, 4) == 'def'
    

def test_not_in_dict_v2():
    assert read_from_dict_v1(test_data, 2) is None


def test_in_dict_v3():
    assert read_from_dict_v1(test_data, 4) == 'def'
    

def test_not_in_dict_v3():
    assert read_from_dict_v1(test_data, 2) is None


In [82]:
!python -m pytest ./dictionary/test_read_from_dict.py

============================= test session starts =============================
platform win32 -- Python 3.7.4, pytest-5.4.3, py-1.9.0, pluggy-0.13.1
rootdir: F:\!EPAM\HW_Lesson12
collected 6 items

dictionary\test_read_from_dict.py ......                                 [100%]

============================== 6 passed in 0.74s ==============================


In [114]:
# Словарь содержит все ключи в диапазоне от 0 до RANGE
def check_read(func):
    for i in range(RANGE):
        temp = func(data, i)

%timeit check_read(read_from_dict_v1)
%timeit check_read(read_from_dict_v2)
%timeit check_read(read_from_dict_v3)

6.16 ms ± 891 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
5.03 ms ± 787 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
5.32 ms ± 906 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [116]:
# Словарь содержит случайный набор ключей 
def check_read(func):
    for i in range(RANGE):
        temp = func(data_random, i)

%timeit check_read(read_from_dict_v1)
%timeit check_read(read_from_dict_v2)
%timeit check_read(read_from_dict_v3)

5.33 ms ± 1.14 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
4.52 ms ± 908 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
6.79 ms ± 1.83 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [117]:
# Словарь содержит все ключи в диапазоне от 0 до RANGE
import cProfile


def benchmark_read(n):
    for _ in range(n):
        for i in range(RANGE):
            read_from_dict_v1(data, i)
            read_from_dict_v2(data, i)
            read_from_dict_v3(data, i)


cProfile.run('benchmark_read(1000)', sort='tottime')

         40000004 function calls in 25.873 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1   12.764   12.764   25.873   25.873 <ipython-input-117-17273a126c3a>:5(benchmark_read)
 10000000    3.807    0.000    5.936    0.000 <ipython-input-102-8dd1d2f31bba>:18(read_from_dict_v1)
 10000000    3.769    0.000    3.769    0.000 <ipython-input-102-8dd1d2f31bba>:26(read_from_dict_v2)
 10000000    3.404    0.000    3.404    0.000 <ipython-input-102-8dd1d2f31bba>:34(read_from_dict_v3)
 10000000    2.129    0.000    2.129    0.000 {method 'get' of 'dict' objects}
        1    0.000    0.000   25.873   25.873 {built-in method builtins.exec}
        1    0.000    0.000   25.873   25.873 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




In [118]:
# Словарь содержит случайный набор ключей 
import cProfile


def benchmark_read_random(n):
    for _ in range(n):
        for i in range(RANGE):
            read_from_dict_v1(data_random, i)
            read_from_dict_v2(data_random, i)
            read_from_dict_v3(data_random, i)


cProfile.run('benchmark_read_random(1000)', sort='tottime')

         40000004 function calls in 28.277 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1   13.070   13.070   28.277   28.277 <ipython-input-118-e1f6cc3e0318>:5(benchmark_read_random)
 10000000    5.141    0.000    5.141    0.000 <ipython-input-102-8dd1d2f31bba>:34(read_from_dict_v3)
 10000000    3.790    0.000    6.824    0.000 <ipython-input-102-8dd1d2f31bba>:18(read_from_dict_v1)
 10000000    3.242    0.000    3.242    0.000 <ipython-input-102-8dd1d2f31bba>:26(read_from_dict_v2)
 10000000    3.035    0.000    3.035    0.000 {method 'get' of 'dict' objects}
        1    0.000    0.000   28.277   28.277 {built-in method builtins.exec}
        1    0.000    0.000   28.277   28.277 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




In [111]:
# Словарь содержит все ключи в диапазоне от 0 до RANGE
%lprun -f benchmark_read benchmark_read(100)

Timer unit: 1e-07 s

Total time: 8.40159 s
File: <ipython-input-60-950e4597868c>
Function: benchmark_read at line 4

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     4                                           def benchmark_read(n):
     5       101       1026.0     10.2      0.0      for _ in range(n):
     6   1000100    9105886.0      9.1     10.8          for i in range(RANGE):
     7   1000000   21790106.0     21.8     25.9              read_from_dict_v1(data, i)
     8   1000000   20656214.0     20.7     24.6              read_from_dict_v2(data, i)
     9   1000000   32462693.0     32.5     38.6              read_from_dict_v3(data, i)

In [ ]:
# %load read_from_dict_memory_profiling.py
from memory_profiler import profile

from dictionary.read_from_dict import read_from_dict_v1, read_from_dict_v2, read_from_dict_v3
from dictionary.read_from_dict import data, RANGE


@profile
def run_all():
    def loop(func):
        for i in range(RANGE):
            temp = func(data, i)

    loop(read_from_dict_v1)
    loop(read_from_dict_v2)
    loop(read_from_dict_v3)


if __name__ == '__main__':
    run_all()


In [73]:
!python read_from_dict_memory_profiling.py

Filename: read_from_dict_memory_profiling.py

Line #    Mem usage    Increment   Line Contents
     7     15.2 MiB     15.2 MiB   @profile
     8                             def run_all():
     9     15.2 MiB      0.0 MiB       def loop(func):
    10     15.2 MiB      0.0 MiB           for i in range(RANGE):
    11     15.2 MiB      0.0 MiB               temp = func(data, i)
    12                             
    13     15.2 MiB      0.0 MiB       loop(read_from_dict_v1)
    14     15.2 MiB      0.0 MiB       loop(read_from_dict_v2)
    15     15.2 MiB      0.0 MiB       loop(read_from_dict_v3)




Замеряла несколько раз. В целом, достаточно ожидаемо, что на случайном наборе ключей функция, где идет обработка прерывания KeyError теряет производительность, в то время как на словаре с полным набором ключей работает примерно на уровне функции с прямым обращением к ключу. А вот почему метод `get` работает чуть медленнее, чем явная проверка на наличие ключа в словаре через `in` и прямым обращением к ключу - это для меня осталось загадкой. Могу только предположить, что на это может влиять внутренняя реализация самого метода `get`.    